File: cli/src/wsc/ops/_chain.py
"""Shared helpers for ops/*.

Every op follows the same pattern:

1. Open an ``audit.start_call`` context with the op name + planned provider.
2. Walk the role's fallback chain. For each provider, try the action lambda.
   On a known provider error, append to ``fallback_chain`` and try the next.
3. On success, mutate the receipt in-place with results/provider/fallback_chain
   and return a payload dict for ``cli._emit``.
"""

from __future__ import annotations

import hashlib
from typing import Any, Callable, Dict, List, Optional, Tuple

from wsc import config as wsc_config
from wsc.providers import (
    AuthError,
    DisabledError,
    MissingKeyError,
    Provider,
    ProviderError,
    RateLimitError,
    TransportError,
    get_provider,
)


def filtered_chain(role: str) -> List[str]:
    """Default chain from config, *not yet* filtered by availability — that's
    decided per-call at provider instantiation time so disable/enable flags
    take effect immediately."""
    return list(wsc_config.ROLE_FALLBACK_CHAIN.get(role, []))


def query_fingerprint(query: str) -> Dict[str, str]:
    if not query:
        return {"query_hash": "", "query_preview": ""}
    h = hashlib.sha256(query.encode("utf-8")).hexdigest()
    return {"query_hash": h, "query_preview": query[:80]}


def run_chain(
    chain: List[str],
    actions: Dict[str, Callable[[Provider], Any]],
) -> Tuple[Optional[str], Any, List[Dict[str, Any]]]:
    """Try each provider in ``chain`` with its action; return
    ``(active_provider, result, fallback_chain)``. ``result`` is whatever the
    action returns (list of NormalizedResult, FetchedPage, etc.) — the op is
    responsible for shaping it."""

    fallback: List[Dict[str, Any]] = []
    for name in chain:
        action = actions.get(name)
        if action is None:
            continue
        try:
            provider = get_provider(name)
        except (MissingKeyError, DisabledError) as exc:
            fallback.append({"from": name, "reason": exc.kind})
            continue
        try:
            result = action(provider)
            return name, result, fallback
        except (RateLimitError, TransportError, AuthError, ProviderError) as exc:
            fallback.append(
                {
                    "from": name,
                    "reason": exc.kind,
                    "error": str(exc)[:200],
                }
            )
            continue
    return None, None, fallback


def chain_failed_payload(op_name: str, fallback: List[Dict[str, Any]]) -> Dict[str, Any]:
    return {
        "ok": False,
        "operation": op_name,
        "provider": None,
        "fallback_chain": fallback,
        "error": f"{op_name}: all providers failed (chain={[f['from'] for f in fallback]})",
        "returncode": 2,
    }

File: src/ops/_chain.ts
/** Shared op helpers — chain runner, fingerprinting, error → fallback step. */

import { ROLE_FALLBACK_CHAIN, type ProviderRole } from "../config.js";
import {
  AuthError,
  DisabledError,
  MissingKeyError,
  Provider,
  ProviderError,
  RateLimitError,
  TransportError,
  getProvider,
} from "../providers/index.js";

export interface FallbackStep {
  from: string;
  to?: string;
  reason: string;
  error?: string;
}

export type Action<R> = (provider: Provider) => R | Promise<R>;

export interface ChainResult<R> {
  active: string | null;
  result: R | null;
  fallback: FallbackStep[];
}

export function filteredChain(role: ProviderRole): string[] {
  return [...(ROLE_FALLBACK_CHAIN[role] ?? [])];
}

export async function runChain<R>(
  chain: string[],
  actions: Record<string, Action<R>>,
): Promise<ChainResult<R>> {
  const fallback: FallbackStep[] = [];
  for (const name of chain) {
    const action = actions[name];
    if (!action) continue;
    let provider: Provider;
    try {
      provider = getProvider(name);
    } catch (err) {
      if (err instanceof MissingKeyError || err instanceof DisabledError) {
        fallback.push({ from: name, reason: err.kind });
        continue;
      }
      throw err;
    }
    try {
      const result = await action(provider);
      return { active: name, result, fallback };
    } catch (err) {
      if (
        err instanceof RateLimitError ||
        err instanceof TransportError ||
        err instanceof AuthError ||
        err instanceof ProviderError
      ) {
        fallback.push({
          from: name,
          reason: err.kind,
          error: String(err.message ?? err).slice(0, 200),
        });
        continue;
      }
      throw err;
    }
  }
  return { active: null, result: null, fallback };
}

export async function chainFailedPayload(
  opName: string,
  fallback: FallbackStep[],
): Promise<{ ok: false; operation: string; provider: null; fallback_chain: FallbackStep[]; error: string; returncode: 2 }> {
  return {
    ok: false,
    operation: opName,
    provider: null,
    fallback_chain: fallback,
    error: `${opName}: all providers failed (chain=${JSON.stringify(fallback.map((f) => f.from))})`,
    returncode: 2,
  };
}

export { queryFingerprint } from "../audit.js";
